crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -313,20 +135,20 @@ impl<'a> Checker<'a> {
 /// Visit an [`Expr`], and treat it as a type definition.
 macro_rules! visit_type_definition {
     ($self:ident, $expr:expr) => {{
-        let prev_in_type_definition = $self.in_type_definition;
-        $self.in_type_definition = true;
+        let prev_in_type_definition = $self.ctx.in_type_definition;
+        $self.ctx.in_type_definition = true;
         $self.visit_expr($expr);
-        $self.in_type_definition = prev_in_type_definition;
+        $self.ctx.in_type_definition = prev_in_type_definition;
     }};
 }
 
 /// Visit an [`Expr`], and treat it as _not_ a type definition.
 macro_rules! visit_non_type_definition {
     ($self:ident, $expr:expr) => {{
-        let prev_in_type_definition = $self.in_type_definition;
-        $self.in_type_definition = false;
+        let prev_in_type_definition = $self.ctx.in_type_definition;
+        $self.ctx.in_type_definition = false;
         $self.visit_expr($expr);
-        $self.in_type_definition = prev_in_type_definition;
+        $self.ctx.in_type_definition = prev_in_type_definition;
     }};
 }
 
```

Comment:
Nit: Should these macros be changed to accept the context as an argument instead to decouple them from `Checker`?

Reply:
They still require `self.visit_expr`, we could change them to accept both though.